sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SdkSpan.java

```diff
@@ -493,12 +515,22 @@ public void end(long timestamp, TimeUnit unit) {
 
   private void endInternal(long endEpochNanos) {
     synchronized (lock) {
-      if (hasEnded) {
-        logger.log(Level.FINE, "Calling end() on an ended Span.");
+      if (hasEnded != EndState.NOT_ENDED) {
+        logger.log(Level.FINE, "Calling end() on an ended or ending Span.");
         return;
       }
       this.endEpochNanos = endEpochNanos;
-      hasEnded = true;
+      spanEndingThread = Thread.currentThread();
```

Comment:
It seems like you could accomplish _almost_ the same thing by just calling `ExtendedSpanProcessor#onEnding` synchronously and setting `hasEnded=true` afterwards, locking it from any asynchronous modification.

But I suppose that if two ExtendedSpanProcessors were registered, one could modify the span asynchronously on a different thread while the second does work synchronously on the calling thread, and the changes could collide.

This mechanism adds additional safety.

Reply:
You mean synchronously while holding the `lock`? Or are you referring to keeping `hasEnded` as a boolean instead of a tri-state?

What I want to protect here against is especially calls from application threads (e.g. `Span.current().setAttribute(..)` modifying the ending span while or after `OnEnding` is executed.

So essentially, one needs three state:
 * The span is freely modifiable by anyone
 * The span is only modifiable by the thread invoking `OnEnding`
 * The span is ended and not modifiable by anyone anymore
 
 One could represent this tri-state by the combination of `boolean hasEnded` and additional on whether `spanEndingThread` is non-null or not, if you prefer that.

Reply:
> Or are you referring to keeping hasEnded as a boolean instead of a tri-state?

This.

> What I want to protect here against is especially calls from application threads (e.g. Span.current().setAttribute(..) modifying the ending span while or after OnEnding is executed.

Yeah I think you've accomplished this in a way that is relatively clean and cheap, so kudos!